kernel/src/expressions/mod.rs

```diff
@@ -209,21 +217,28 @@ impl Display for Expression {
 
 impl Expression {
     /// Returns a set of columns referenced by this expression.
-    pub fn references(&self) -> HashSet<&str> {
+    pub fn references(&self) -> HashSet<&ColumnName> {
         let mut set = HashSet::new();
 
         for expr in self.walk() {
             if let Self::Column(name) = expr {
-                set.insert(name.as_str());
+                set.insert(name);
             }
         }
 
         set
     }
 
-    /// Create an new expression for a column reference
-    pub fn column(name: impl ToString) -> Self {
-        Self::Column(name.to_string())
+    /// Creates a new column reference, for a simple (not nested) column name.
+    /// See [`ColumnName::simple`] for details.
+    pub fn simple_column(name: impl Into<String>) -> Self {
+        ColumnName::simple(name).into()
+    }
+
+    /// Creates a new column reference, for a splittable nested column name.
+    /// See [`ColumnName::split`] for details.
+    pub fn split_column(name: impl AsRef<str>) -> Self {
```

Comment:
Seems that I need to know a priori whether a column name is nested or not when calling  `Expression::simple_column` vs `Expression::split_column`. Do we want devs to have to keep track of that? Why not have a  `column(name: impl AsRef<str>)`, and the function can figure out if it's split or simple?

Reply:
I went ahead and defined macros that give provable compile-time safety

Reply:
awesome i really like the macros :)

Reply:
at risk of a dumb question: could we just have a `simple_column` be something like: either a nested column split by periods (and no periods in field names) like `a.b.c` or just a single top-level column? then we wouldn't need two APIs for top-level/nested column names? It could just unify under `simple_column!("nested.col") -> ColumnName::new([name, col])` and `simple_column!("not_nested") -> ColumnName::new([not_nested])` or something like that?

Reply:
and, come to think of it, regarding naming: I wonder if just something like `column_name!()` macro would suffice, since it's compile-time checked if any error occurs we can point back to macro docs and have a clear error message that says 'column_name!' is for generating a simple '`.`-delimited' column name without special chars, etc, etc

Reply:
Woah that was an excellent overview of the problems of parsing and how this addresses them. Thank you!

Makes sense that we're avoiding parsing in the kernel. We're at a low enough level that any names we're passed are already parsed.

I'll take a look at the macros this review. Excited to see what you cooked up 👀

Reply:
> I wonder if just something like column_name!() macro would suffice, since it's compile-time checked if any error occurs we can point back to macro docs and have a clear error message that says 'column_name!' is for generating a simple '.-delimited' column name without special chars, etc, etc

That's a great question. I went back and forth quite a bit and settled on the current form because it requires developer to state intent and allows code reviewers to easily validate intent (e.g. `simple_column!` blows up if it sees any dots. I even debated having `nested_column!` blow up _unless_ it sees at least one dot, but that seemed overkill. 

That said, I'm very open to discussion if we think the current split is also overkill. Maybe we're already safe enough from the combination of string literals and the fact that Delta spec column names never contain dots.

Reply:
Thinking more, I agree that two sets of macros is overkill. Simplified to just one concept: `column_name!` for a `ColumnName` and `column_expr!` for a `Expression::Column`.